File: core/core-api/src/main/java/com/coffee/gu/cancel/CancelEventDispatcher.java
package com.coffee.gu.cancel;

import com.coffee.gu.CancelEvent;
import com.coffee.gu.EventDispatcher;
import com.coffee.gu.enums.EventType;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Component;
import tools.jackson.databind.json.JsonMapper;

@Component
public class CancelEventDispatcher implements EventDispatcher {

    private final JsonMapper jsonMapper;
    private final ApplicationEventPublisher applicationEventPublisher;

    public CancelEventDispatcher(JsonMapper jsonMapper, ApplicationEventPublisher applicationEventPublisher) {
        this.jsonMapper = jsonMapper;
        this.applicationEventPublisher = applicationEventPublisher;
    }

    @Override
    public boolean supports(EventType eventType) {
        return eventType == EventType.CANCEL;
    }

    @Override
    public void dispatch(String payload) {
        CancelEvent cancelEvent = jsonMapper.readValue(payload, CancelEvent.class);
        applicationEventPublisher.publishEvent(cancelEvent);
    }
}

File: core/core-api/src/main/kotlin/com/coffee/gu/cancel/CancelEventDispatcher.kt
package com.coffee.gu.cancel

import com.coffee.gu.CancelEvent
import com.coffee.gu.EventDispatcher
import com.coffee.gu.enums.EventType
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Component
import tools.jackson.databind.json.JsonMapper

@Component
class CancelEventDispatcher(
    private val jsonMapper: JsonMapper,
    private val applicationEventPublisher: ApplicationEventPublisher,
) : EventDispatcher {

    override fun supports(eventType: EventType): Boolean {
        return eventType == EventType.CANCEL
    }

    override fun dispatch(payload: String) {
        val cancelEvent = jsonMapper.readValue(payload, CancelEvent::class.java)
        applicationEventPublisher.publishEvent(cancelEvent)
    }
}
